### PR TITLE
fix wrong docs

### PR DESCRIPTION
**Pull request by StevenPuttemans**
It is simply unlogical that the overall hit rate of a cascade would be worse then the individual minimal hit rates for each seperate weak classifier. I guess this is a copy paste error from the following feature that can be set, being the maxFalseAlarmRate.

Example:
A cascade with `-minHitRate 0.95` and `-numStages 20`, would according to this function, have a overall hit rate of 0.35, which is terribly low. 

In my opinion this cascade would have a minHitRate of at least 0.95, and even higher if the individual weak classifiers have higher values. How they combine correctly, I will investigate and push in another PR.

### CODE REVIEW

**Comment by alalek** [comment]
I believe that
```
overall_hit_rate = 1.0 - overall_fail_rate
overall_fail_rate = max_fail_rate^number_of_stages
max_fail_rate = 1.0 - min_hit_rate
```

Totally,
`overall_hit_rate = (1.0 - (1.0 - min_hit_rate)^number_of_stages)`

/cc @vpisarev

**Comment by StevenPuttemans** [comment]
Well lets imagine that the following 5 stage cascade with a minHitRate of 0.95 was generated, that would lead to a overall hit rate = ( 1.0 - (1.0 - 0.95)^5) = 0.999999688, which means that about 4 in 10.000.000 positives would classify as negatives, which is actually not bad on a training set and more acceptable in reasoning.

I will add the formula!

**Comment by teng88** [comment]
The overall hit rate in theory should be: min_hit_rate^number_of_stages.
The point is that in each stage a threshold is selected just to let min_hit_rate percent of all passed positives pass this stage, so the overall hit rate for the whole classifier is min_hit_rate^number_of_stages.
In actual implementation, since the actual hit rate of each stage >= min_hit_rate (because sometimes the threshold happen to be equal to the feature value of the positives preparing to reject), the formula can be more precise as actual_hit_rate[0]*actual_hit_rate[1]*...

**Comment by StevenPuttemans** [comment]
_The point is that in each stage a threshold is selected just to let min_hit_rate percent of all passed positives pass this stage, so the overall hit rate for the whole classifier is min_hit_rate^number_of_stages._

Uhm wait, this cannot be true, as far as I understand the theory correctly, `minimum hit rate` means the amount of positive samples that need to be classified as positives for that specific stage. Setting it at 0.95 means that 5 out of 100 positive samples of that stage can be wrongly classified.

So it would be crazy that the overall hit rate would be lower then the min hit rate ...

**Comment by teng88** [comment]
As in your case, e.g., 5 are rejected and 95 samples are passed for the training in the 2nd stage, the threshold will be selected  just to let 95% percent of the these 95 samples to pass this stage.

**Comment by StevenPuttemans** [comment]
> As in your case, e.g., 5 are rejected and 95 samples are passed for the training in the 2nd stage, the threshold will be selected just to let 95% percent of the these 95 samples to pass this stage.

That is not true. If 5 of the 100 samples are rejected, 5 new samples will be grabbed from the positive vec file, to start the training of the 2nd stage again with 100 positive training samples. This is the exact reason why you need to pass a lower value to `-numPos` than the actual amount of samples in the vec file.

If your reasoning is followed, then only +- 35% of the initial amount of positive samples will be used to train the later stages, which is simply incorrect.

**Comment by teng88** [comment]
> That is not true. If 5 of the 100 samples are rejected, 5 new samples will be grabbed from the positive vec file, to start the training of the 2nd stage again with 100 positive training samples. This is the exact reason why you need to pass a lower value to -numPos than the actual amount of samples in the vec file.

You are right. Assume that exact 5 positives are reject for each stage, then the final detection rate can be computed as passed_positives / all_used_positives = 95 / (95 + 5*20), right?

**Comment by StevenPuttemans** [comment]
> ... then the final detection rate for the whole classifier can be computed as passed_positives / all_used_positives = 95 / (95 + 5*20) ...

I am afraid that it is not that easy. The hit rate ONLY tells you stuff on the training data, not on the detection capabilities of the trained cascade, because those depend on the classification of a test set. For example, it is perfectly possible to train a classifier with a detection rate of nearly 100% on the positive training data, that only works for 75% on the test data, simply because some knowledge from the test data was not contained inside the training data. Also the hit rate tells you nothing about negative sample classification and so the step directly to classifier detection rate is just to easy.

**Comment by teng88** [comment]
> I am afraid that it is not that easy. The hit rate ONLY tells you stuff on the training data, not on the detection capabilities of the trained cascade, because those depend on the classification of a test set. For example, it is perfectly possible to train a classifier with a detection rate of nearly 100% on the positive training data, that only works for 75% on the test data, simply because some knowledge from the test data was not contained inside the training data. 

I agree with you in that point, the formula is only an approximation for trainging data. In my experience, the detection rate usually drop a certain extent for test dataset.

**Comment by StevenPuttemans** [comment]
@alalek the packing warning, do you have any idea what it means?

**Comment by StevenPuttemans** [comment]
@vpisarev or @alalek might be a chance that this can get merged in? trying to clean everything up here locally :)

**Comment by catree** [comment]
I am also interested in this discussion even if I am not at all an expert in this field.
I believe the original formula `overall_hit_rate = min_hit_rate^number_of_stages` is correct and I will try to add the information that make me think that.

My answer to the original question: 

> It is simply unlogical that the overall hit rate of a cascade would be worse then the individual minimal hit rates for each seperate weak classifier

What I have understood so far is that the principle of the cascade of classifiers is:

- the first stage of the cascade is built in a way to detect all the positive examples whereas rejecting a significant number of negatives examples (50%). It is often a rather simple classifier with few features.

- but it means also that 50% of the negative examples are classified as positive (false positives).

- the succession of stages in the cascade will little by little rejects the negative examples by using more and more complex classifiers: a classifier in a stage is trained on negative examples that passed the previous stage (false positives) and thus often more features are needed to be able to discriminate the example.

- this mechanism (more features / more complex classifiers when we go deeper in the cascade) combined with the usage of stages are the key to reduce drastically the number of false positives but in the same time will impact the true positive rate (or hit rate) of the overall cascade of classifiers. 

- so if this mechanism allows to reduce the number of false positives of the cascade as the product of the false positive rate of each stage, the same think should apply for the overall detection rate.

- in fact it is easy to build a classifier that has a good hit rate if you accept to have lots of false positives. To reduce the number of false positives, you will have to constraint more the feature space but mechanically you will reduce slightly the hit rate.

What I have written may be obvious for you. So, by looking quickly in the litterature, what I have found seems to confirm the original formula:

- [Rapid object detection using a boosted cascade of simple features, Viola & Jones, 2001](https://www.cs.cmu.edu/~efros/courses/LBMV07/Papers/viola-cvpr-01.pdf):

> Each stage in the cascade reduces the false positive rate and decreases the detection rate. A target is selected for the minimum reduction in false positives and the maximum decrease in detection. Each stage is trained by adding features until the target detection and false positives rates are met (these rates are determined by testing the detector on a validation set). Stages are added until the overall target for false positive and detection rate is met.

- [Robust Real-Time Face Detection, Viola & Jones, 2004](http://www.vision.caltech.edu/html-files/EE148-2005-Spring/pprs/viola04ijcv.pdf):

> For example a detection rate of 0.9 can be achieved by a 10 stage classifier if each stage has a detection rate of 0.99 (since 0.9 ≈ 0.99^10). While achieving this detection rate may sound like a daunting task, it is made significantly easier by the fact that each stage need only achieve a false positive rate of about 30% (0.30^10 ≈ 6 × 10−6).

The explicit formula also appears in §4.1 and in table 2.

- [Robust Object Detection Via Soft Cascade, Bourdev & Brandt, 2005](http://www.lubomir.org/pdfs/softcascade.pdf):

> A second disadvantage of the cascade structure is the severe requirement it imposes on training each of the individual stages. In particular, a positive classification must pass every stage, which implies that the final detection rate is the product of the detection rates of all stages. For a 10 stage cascade to achieve a detection rate of 90% at false positive rate of 10−6, each of its stages must be able to preserve close to 99% of the faces, while eliminating on average more than 25% of the non-faces. This may be trivial  for the initial stages, but for later stages it becomes a very difficult task, which results in large stages that are difficult to train and time-consuming to evaluate.

- The same formula / reasoning appears in this [course](https://www.cl.cam.ac.uk/teaching/1516/CompVision/slides2016.pdf) page 108 by Professor John Daugman.

Also, if I follow your formula: 

> overall hit rate = (1.0 - (1.0 - 0.95)^5) = 0.999999688

for a 20-stages cascade of classifers it produces the overall detection rate: `(1.0 - (1.0 - 0.95)^20) = 0.99999999999999999999999999046326` or `9.5367431640625e-27` false negatives (or miss). Imagine if we have trained with a stage hit rate of 0.999!!!

Sorry for this long post but I wanted to be clear as much as possible.

Also, I used this Wikipedia page on [Precision and recall](https://en.wikipedia.org/wiki/Precision_and_recall):

- for me I understand: detection rate = true positive rate = hit rate = TP / (TP + FN)

- and false positive rate = FP / (FP + TN)

**Comment by StevenPuttemans** [comment]
First of all, thank you for this very nice explanation. It is nice to see that some people are willing to cooperate in finding the correct term.

I kind of agree with everything you are stating above, and thus am feeling that indeed, the formula could be correct, but bare with me for one more rant.

If you consider all your input, then I agree that increasing the drop in false positives, will result in true positives being dropped at some point. You have indeed a tradeoff and thus something has to go. BUT, in the case of the default value of 0.95, that would mean that the first model is allowed to wrongly classify 5 of 100 positive samples wrong. BUT, by stage 20, it would reach a value of 0.35 and thus a model that is only able to correctly classify 35 procent, so basically 65 out of 100 samples do not get through ...

I would agree if this is the hit rate of the last stage only, but overall? That would mean that training is stupid, because we will generate TONS of false negatives by simply making the model better ...

**Comment by alalek** [comment]
@catree Thank you for great explanation.

I believe cascade scheme for true positives processing is (assume all stages have similar "parameters"):
```
stage 0 "true positives":
< ++++++++++++++++++++++++++++++++++++++++++++++++++ | +++++ >
                   future processing                 |  drop
stage 1:
< +++++++++++++++++++++++++++++++++++++++++++ | ++++ >
                future processing             | drop
stage 2:
< ++++++++++++++++++++++++++++++++++++ | ++++ >
             future processing         | drop
stage 3:
< +++++++++++++++++++++++++++++ | ++++ >
         future processing      | drop
...
stage N:
< ++ >
```

For each cascade stage it is allowed to drop (filter out) up to (1-minHitRate) positives (with many negatives, which are not showed in scheme).

@StevenPuttemans I believe it will be good to add this link into documentation:
§4.1 [Robust Real-Time Face Detection, Viola & Jones, 2004](http://www.vision.caltech.edu/html-files/EE148-2005-Spring/pprs/viola04ijcv.pdf)

**Comment by StevenPuttemans** [comment]
I guess this all works and falls together then. I will indeed add the documentation link when ready. As to the test sample, simply forget it. I am noticing that my whole point was built on 0.95 as minhitrate, while default is 0.995 ... thats makes much more sense, since a 20 level classifier still finds 90% of all detected objects. I guess that is more than achievable. I shall reopen a new PR later to add the documentation, or someone with more time now can feel free to do so!

**Comment by catree** [comment]
Thank you for the quick review.
I am happy if I can help a little considering the large benefit I gain for having OpenCV and for free.
